scalikejdbc-core/src/main/scala/scalikejdbc/DBSession.scala

```diff
@@ -617,40 +617,40 @@ trait DBSession extends LogSupport with LoanPattern with AutoCloseable {
    *
    * @param template SQL template
    * @param params parameters
-   * @return generated key as a long value
+   * @return generated key as a T value
    */
-  def updateAndReturnGeneratedKey(template: String, params: Any*): Long = updateAndReturnSpecifiedGeneratedKey(template, params: _*)(1)
+  def updateAndReturnGeneratedKey[T](template: String, params: Any*): T = updateAndReturnSpecifiedGeneratedKey[T](template, params: _*)(1)
 
   /**
    * Executes java.sql.PreparedStatement#executeUpdate() and returns the generated key.
    *
    * @param template SQL template
    * @param params parameters
    * @param key name
-   * @return generated key as a long value
+   * @return generated key as a T value
    */
-  def updateAndReturnSpecifiedGeneratedKey(template: String, params: Any*)(key: Any): Long = {
+  def updateAndReturnSpecifiedGeneratedKey[T](template: String, params: Any*)(key: Any): T = {
     var generatedKeyFound = false
-    var generatedKey: Long = -1
+    var generatedKey: T = null.asInstanceOf[T]
     val before = (stmt: PreparedStatement) => {}
     val after = (stmt: PreparedStatement) => {
       val rs = stmt.getGeneratedKeys
       while (rs.next()) {
         generatedKeyFound = true
         generatedKey = key match {
           case name: String => try {
-            rs.getLong(name)
+            rs.getObject(name).asInstanceOf[T]
```

Comment:
I am not sure if using getObject and casting the value is the best way yet. I would like to know what other libraries do for this purpose. Do you have any pieces of evidence for it?

Reply:
@stefanbacon 

```scala
rs.get[T](name)
```

Maybe this works?